diesel/src/pg/connection/cursor.rs

```diff
@@ -66,112 +57,348 @@ impl Iterator for Cursor<'_> {
     }
 }
 
-#[test]
-fn fun_with_row_iters() {
-    crate::table! {
-        #[allow(unused_parens)]
-        users(id) {
-            id -> Integer,
-            name -> Text,
+/// The type returned by various [`Connection`] methods.
+/// Acts as an iterator over `T`.
+#[allow(missing_debug_implementations)]
+pub struct RowByRowCursor<'a> {
+    current_row: usize,
+    db_result: Rc<PgResult>,
+    conn: &'a mut super::ConnectionAndTransactionManager,
+}
+
+impl<'a> RowByRowCursor<'a> {
+    pub(super) fn new(
+        db_result: PgResult,
+        conn: &'a mut super::ConnectionAndTransactionManager,
+    ) -> Self {
+        RowByRowCursor {
+            current_row: 0,
+            db_result: Rc::new(db_result),
+            conn,
         }
     }
+}
 
-    use crate::deserialize::{FromSql, FromSqlRow};
-    use crate::pg::Pg;
-    use crate::prelude::*;
-    use crate::row::{Field, Row};
-    use crate::sql_types;
-
-    let conn = &mut crate::test_helpers::connection();
-
-    crate::sql_query(
-        "CREATE TABLE IF NOT EXISTS users(id INTEGER PRIMARY KEY, name TEXT NOT NULL);",
-    )
-    .execute(conn)
-    .unwrap();
-
-    crate::insert_into(users::table)
-        .values(vec![
-            (users::id.eq(1), users::name.eq("Sean")),
-            (users::id.eq(2), users::name.eq("Tess")),
-        ])
+impl Iterator for RowByRowCursor<'_> {
+    type Item = crate::QueryResult<PgRow>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current_row > 0 {
+            let get_next_result = super::update_transaction_manager_status(
+                self.conn.raw_connection.get_next_result(),
+                self.conn,
+            );
+            match get_next_result {
+                Ok(Some(res)) => {
+                    if let Some(old_res) = Rc::get_mut(&mut self.db_result) {
+                        *old_res = res;
+                    } else {
+                        self.db_result = Rc::new(res);
+                    }
+                    self.current_row = 0;
+                }
+                Ok(None) => {
+                    return None;
+                }
+                Err(e) => return Some(Err(e)),
+            }
+        }
+        if self.current_row < self.db_result.num_rows() {
+            let row = self.db_result.clone().get_row(self.current_row);
+            self.current_row += 1;
+            Some(Ok(row))
+        } else {
+            None
+        }
+    }
+}
+
+impl Drop for RowByRowCursor<'_> {
+    fn drop(&mut self) {
+        loop {
```

Comment:
Is there no way to cancel the rest of the execution of the query and not load any more?

Reply:
https://www.postgresql.org/docs/13/libpq-single-row-mode.html states that you have to call `get_next_result()` till it returns a null pointer. 

The only way to cancel server side operations is:
https://www.postgresql.org/docs/current/libpq-cancel.html
If I read that correctly, this pages says that this operation is best effort and may change nothing. This would imply that we need to call get_next_result()` in a loop nevertheless just to be sure.

Reply:
That is probably desirable behavior still if the iterator is not already entirely consumed: if we are looping on a large dataset (which will generally be the use-case for row-by-row loading) and break in the middle, I think we want to tell the database that it doesn't need to send the entire rest of its results: If it is large, data is not already on the client, so we can try to prevent the database from sending them, or even compute them.

Reply:
https://stackoverflow.com/questions/18015203/postgresql-cancel-query-from-c-c-program indicates that canceling such queries won't do much.